File: src/main/java/net/indevo/fantasy_metals/datagen/ModBlockTagGenerator.java
package net.indevo.fantasy_metals.datagen;

import net.indevo.fantasy_metals.FantasyMetals;
import net.indevo.fantasy_metals.block.ModBlocks;
import net.indevo.fantasy_metals.util.ModTags;
import net.minecraft.core.HolderLookup;
import net.minecraft.core.registries.Registries;
import net.minecraft.data.PackOutput;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.tags.BlockTags;
import net.minecraft.tags.TagKey;
import net.minecraft.world.level.block.Block;
import net.minecraftforge.common.data.BlockTagsProvider;
import net.minecraftforge.common.data.ExistingFileHelper;
import org.jetbrains.annotations.Nullable;

import java.util.concurrent.CompletableFuture;

public class ModBlockTagGenerator extends BlockTagsProvider {
    public ModBlockTagGenerator(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, @Nullable ExistingFileHelper existingFileHelper) {
        super(output, lookupProvider, FantasyMetals.MOD_ID, existingFileHelper);
    }

    private static TagKey<Block> create(String p_203847_) {
        return TagKey.create(Registries.BLOCK, new ResourceLocation(p_203847_));
    }

    @Override
    protected void addTags(HolderLookup.Provider pProvider) {
        this.tag(BlockTags.MINEABLE_WITH_PICKAXE)
                .add(
                        ModBlocks.MYTHRIL_BLOCK.get(),
                        ModBlocks.RAW_MYTHRIL_BLOCK.get(),
                        ModBlocks.MYTHRIL_ORE.get(),
                        ModBlocks.DEEPSLATE_MYTHRIL_ORE.get(),

                        ModBlocks.ADAMANT_BLOCK.get(),
                        ModBlocks.RAW_ADAMANT_BLOCK.get(),
                        ModBlocks.ADAMANT_ORE.get(),

                        ModBlocks.CARMOT_BLOCK.get(),
                        ModBlocks.RAW_CARMOT_BLOCK.get(),
                        ModBlocks.CARMOT_ORE.get(),
                        ModBlocks.DEEPSlATE_CARMOT_ORE.get(),

                        ModBlocks.SARDONYX_ORE.get(),
                        ModBlocks.DEEPSLATE_SARDONYX_ORE.get(),

                        ModBlocks.ALEXANDRITE_ORE.get(),
                        ModBlocks.DEEPSLATE_ALEXANDRITE_ORE.get(),

                        ModBlocks.SARDONYX_BLOCK.get(),
                        ModBlocks.ALEXANDRITE_BLOCK.get(),
                        ModBlocks.TSAVORITE_BLOCK.get(),
                        ModBlocks.TANZANITE_BLOCK.get(),
                        ModBlocks.BLACK_OPAL_BLOCK.get(),

                        ModBlocks.TANZANITE_ORE.get(),

                        ModBlocks.BLACK_OPAL_ORE.get(),

                        ModBlocks.TSAVORITE_ORE.get(),
                        ModBlocks.DEEPSLATE_TSAVORITE_ORE.get(),

                        ModBlocks.ORICHALCUM_BLOCK.get(),
                        ModBlocks.RAW_ORICHALCUM_BLOCK.get(),
                        ModBlocks.ORICHALCUM_ORE.get(),
                        ModBlocks.DEEPSLATE_ORICHALCUM_ORE.get());

        this.tag(BlockTags.NEEDS_IRON_TOOL)
                .add(
                        ModBlocks.ORICHALCUM_BLOCK.get(),
                        ModBlocks.RAW_ORICHALCUM_BLOCK.get(),
                        ModBlocks.ORICHALCUM_ORE.get(),

                        ModBlocks.CARMOT_BLOCK.get(),
                        ModBlocks.RAW_CARMOT_BLOCK.get(),
                        ModBlocks.CARMOT_ORE.get(),
                        ModBlocks.DEEPSlATE_CARMOT_ORE.get(),

                        ModBlocks.BLACK_OPAL_ORE.get(),
                        ModBlocks.BLACK_OPAL_BLOCK.get(),

                        ModBlocks.TANZANITE_ORE.get(),
                        ModBlocks.TANZANITE_BLOCK.get(),

                        ModBlocks.SARDONYX_ORE.get(),
                        ModBlocks.DEEPSLATE_SARDONYX_ORE.get(),
                        ModBlocks.SARDONYX_BLOCK.get(),

                        ModBlocks.ALEXANDRITE_ORE.get(),
                        ModBlocks.DEEPSLATE_ALEXANDRITE_ORE.get(),
                        ModBlocks.ALEXANDRITE_BLOCK.get(),

                        ModBlocks.DEEPSLATE_ORICHALCUM_ORE.get());

        this.tag(BlockTags.NEEDS_DIAMOND_TOOL)
                .add(
                        ModBlocks.MYTHRIL_BLOCK.get(),
                        ModBlocks.RAW_MYTHRIL_BLOCK.get(),
                        ModBlocks.MYTHRIL_ORE.get(),
                        ModBlocks.DEEPSLATE_MYTHRIL_ORE.get(),

                        ModBlocks.TSAVORITE_ORE.get(),
                        ModBlocks.DEEPSLATE_TSAVORITE_ORE.get(),
                        ModBlocks.TSAVORITE_BLOCK.get(),

                        ModBlocks.ADAMANT_BLOCK.get(),
                        ModBlocks.RAW_ADAMANT_BLOCK.get(),
                        ModBlocks.ADAMANT_ORE.get());

        this.tag(ModTags.Blocks.STORAGE_BLOCKS)
                .add(
                        ModBlocks.MYTHRIL_BLOCK.get(),
                        ModBlocks.RAW_MYTHRIL_BLOCK.get(),

                        ModBlocks.ORICHALCUM_BLOCK.get(),
                        ModBlocks.RAW_ORICHALCUM_BLOCK.get(),

                        ModBlocks.CARMOT_BLOCK.get(),
                        ModBlocks.RAW_CARMOT_BLOCK.get(),

                        ModBlocks.SARDONYX_BLOCK.get(),
                        ModBlocks.ALEXANDRITE_BLOCK.get(),
                        ModBlocks.TSAVORITE_BLOCK.get(),
                        ModBlocks.TANZANITE_BLOCK.get(),
                        ModBlocks.BLACK_OPAL_BLOCK.get(),

                        ModBlocks.ADAMANT_BLOCK.get(),
                        ModBlocks.RAW_ADAMANT_BLOCK.get());

        this.tag(ModTags.Blocks.ORES)
                .add(
                        ModBlocks.MYTHRIL_ORE.get(),
                        ModBlocks.DEEPSLATE_MYTHRIL_ORE.get(),

                        ModBlocks.ORICHALCUM_ORE.get(),
                        ModBlocks.DEEPSLATE_ORICHALCUM_ORE.get(),

                        ModBlocks.SARDONYX_ORE.get(),
                        ModBlocks.DEEPSLATE_SARDONYX_ORE.get(),

                        ModBlocks.ALEXANDRITE_ORE.get(),
                        ModBlocks.DEEPSLATE_ALEXANDRITE_ORE.get(),

                        ModBlocks.TANZANITE_ORE.get(),

                        ModBlocks.BLACK_OPAL_ORE.get(),

                        ModBlocks.TSAVORITE_ORE.get(),
                        ModBlocks.DEEPSLATE_TSAVORITE_ORE.get(),

                        ModBlocks.CARMOT_ORE.get(),
                        ModBlocks.DEEPSlATE_CARMOT_ORE.get(),

                        ModBlocks.ADAMANT_ORE.get());
    }
}

File: src/main/kotlin/net/teamsolar/fantasy_metals/datagen/ModBlockTagGenerator.kt
package net.teamsolar.fantasy_metals.datagen

import net.minecraft.core.HolderLookup
import net.minecraft.data.PackOutput
import net.minecraft.data.tags.IntrinsicHolderTagsProvider
import net.minecraft.resources.ResourceLocation
import net.minecraft.tags.BlockTags
import net.minecraft.tags.TagKey
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.Blocks
import net.neoforged.neoforge.common.Tags
import net.neoforged.neoforge.common.data.BlockTagsProvider
import net.neoforged.neoforge.common.data.ExistingFileHelper
import net.neoforged.neoforge.registries.DeferredBlock
import net.teamsolar.fantasy_metals.FantasyMetals
import net.teamsolar.fantasy_metals.ModTags
import net.teamsolar.fantasy_metals.block.ModBlocks
import net.teamsolar.fantasy_metals.item.GemItemSet
import net.teamsolar.fantasy_metals.item.ItemSet
import net.teamsolar.fantasy_metals.item.MetalEquipmentSet
import net.teamsolar.fantasy_metals.item.ModItems
import java.util.concurrent.CompletableFuture

class ModBlockTagGenerator(
    output: PackOutput, lookupProvider: CompletableFuture<HolderLookup.Provider>,
    existingFileHelper: ExistingFileHelper
) :
    BlockTagsProvider(output, lookupProvider, FantasyMetals.MODID, existingFileHelper) {
    override fun addTags(provider: HolderLookup.Provider) {
        val providerHelper = object: BlockTagsProviderHelper, ModBlockInputHelper {
            override fun tag(tagKey: TagKey<Block>): IntrinsicTagAppender<Block> = this@ModBlockTagGenerator.tag(tagKey)
            override val ORE_BLOCKS: List<DeferredBlock<out Block>> = ModBlocks.ORE_BLOCKS
            override val RAW_BLOCKS: List<DeferredBlock<out Block>> = ModBlocks.RAW_BLOCKS
            override val METAL_BLOCKS: List<DeferredBlock<out Block>> = ModBlocks.METAL_BLOCKS
        }
        fun MetalEquipmentSet.thenAddTags() = providerHelper.addTags()
        fun GemItemSet.thenAddTags() = providerHelper.addTags()
        fun IntrinsicHolderTagsProvider.IntrinsicTagAppender<Block>.addSet(set: ItemSet) {
            for(block in providerHelper.blocksWithPrefix(set.prefix)) {
                add(block.get())
            }
        }

        tag(ModTags.Blocks.INCORRECT_FOR_MYTHRIL_TOOL).addTag(BlockTags.INCORRECT_FOR_DIAMOND_TOOL)
        tag(ModTags.Blocks.INCORRECT_FOR_ADAMANT_TOOL).addTag(BlockTags.INCORRECT_FOR_NETHERITE_TOOL)
        tag(ModTags.Blocks.INCORRECT_FOR_ORICHALCUM_TOOL).addTag(BlockTags.INCORRECT_FOR_IRON_TOOL)
        tag(ModTags.Blocks.INCORRECT_FOR_CARMOT_TOOL).addTag(BlockTags.INCORRECT_FOR_DIAMOND_TOOL)

        ModItems.MYTHRIL_SET.thenAddTags()
        tag(BlockTags.NEEDS_DIAMOND_TOOL).addSet(ModItems.MYTHRIL_SET)
        ModItems.ADAMANT_SET.thenAddTags()
        tag(BlockTags.NEEDS_DIAMOND_TOOL).addSet(ModItems.ADAMANT_SET)
        ModItems.ORICHALCUM_SET.thenAddTags()
        tag(BlockTags.NEEDS_IRON_TOOL).addSet(ModItems.ORICHALCUM_SET)
        ModItems.CARMOT_SET.thenAddTags()
        tag(BlockTags.NEEDS_IRON_TOOL).addSet(ModItems.CARMOT_SET)


        ModItems.ALEXANDRITE_SET.thenAddTags()
        tag(BlockTags.NEEDS_IRON_TOOL).addSet(ModItems.ALEXANDRITE_SET)
        ModItems.BLACK_OPAL_SET.thenAddTags()
        tag(BlockTags.NEEDS_IRON_TOOL).addSet(ModItems.BLACK_OPAL_SET)
        ModItems.TANZANITE_SET.thenAddTags()
        tag(BlockTags.NEEDS_IRON_TOOL).addSet(ModItems.TANZANITE_SET)
        ModItems.SARDONYX_SET.thenAddTags()
        tag(BlockTags.NEEDS_IRON_TOOL).addSet(ModItems.SARDONYX_SET)
        ModItems.TSAVORITE_SET.thenAddTags()
        tag(BlockTags.NEEDS_IRON_TOOL).addSet(ModItems.TSAVORITE_SET)
        // tag(BlockTags.NEEDS_IRON_TOOL).addSet(ModItems.)

        tag(Tags.Blocks.ORES_IN_GROUND_STONE).add(
            ModBlocks.MYTHRIL_ORE.get(),
            ModBlocks.ORICHALCUM_ORE.get(),
            ModBlocks.CARMOT_ORE.get(),

            ModBlocks.SARDONYX_ORE.get(),
            ModBlocks.ALEXANDRITE_ORE.get(),
            ModBlocks.TSAVORITE_ORE.get()
        )
        tag(Tags.Blocks.ORES_IN_GROUND_NETHERRACK).add(
            ModBlocks.BLACK_OPAL_ORE.get()
        )
        tag(ModTags.Blocks.ORES_IN_GROUND_ENDSTONE).add(
            ModBlocks.ADAMANT_ORE.get(),
            ModBlocks.TANZANITE_ORE.get()
        )
        tag(ModTags.Blocks.ORE_BEARING_ENDSTONE).add(
            Blocks.END_STONE
        )
    }

    override fun getName(): String {
        return "Block Tags"
    }
}
